Named Services

Br4nd0R opened the pull request:
# Changes
* CumulusCI now supports named services! This means you can configure multiple services of the same _type_ under different names. If you run `cci service list` you will note that your existing global services will have the name `service_type__global`, and any project specific services will have the name `service_type__project_name`.
* CumulusCI has a new command: `cci service default`. This command sets the default service for a given type.
* CumulusCI has a new command: `cci service rename`. This command renames a given service.

## Internal Considerations
* `keychain.list_services()` now returns a dict with service types as keys, and each key returns a list of available aliases/names of services configured for the given type. (I believe this has potential impacts on MetaCI).
* The keychain has a new public API method `set_default_service(service_type, alias)` for setting the default service for a given type. (This will be useful for the new `cci service default` command).
* The keychain's `get_service()` method will return the the default service for the given type if no alias is specified.

## Misc
* The `make coverage` target now supports a `CLASS_PATH` argument to allow running coverage for a subset of tests.

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -1,16 +1,40 @@
+Service files (.service) are organized under ~/.cumulusci/services/ into sub-directories that
+pertain to a particular service type (e.g. github, connected_app). These files hold
+the encrypted 'attributes' of each service as they are listed in the universal
+cumulusci.yml file. There are multiple DEFAULT_SERVICES.json files, the map a particular
```
```suggestion
cumulusci.yml file. There are multiple DEFAULT_SERVICES.json files that map a particular
```
```suggestion
cumulusci.yml file. There are multiple DEFAULT_SERVICES.json files, the map a particular
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -1,16 +1,40 @@
+cumulusci.yml file. There are multiple DEFAULT_SERVICES.json files, the map a particular
+service type to the default alias for that taype.
+
+The DEFUALT_SERVICES.json file that resides at ~/.cumulusci holds the global default
```
```suggestion
The DEFAULT_SERVICES.json file that resides at ~/.cumulusci holds the global default
```
```suggestion
The DEFUALT_SERVICES.json file that resides at ~/.cumulusci holds the global default
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -1,16 +1,40 @@
+
+The DEFUALT_SERVICES.json file that resides at ~/.cumulusci holds the global default
+services mappings. Each local project directory (~/.cumulusci/project/) also has a
+DEFUALT_SERVICES.json file for any services that have been configured to be project specific.
```
```suggestion
DEFAULT_SERVICES.json file for any services that have been configured to be project specific.
```
```suggestion
DEFUALT_SERVICES.json file for any services that have been configured to be project specific.
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -1,16 +1,40 @@
+services mappings. Each local project directory (~/.cumulusci/project/) also has a
+DEFUALT_SERVICES.json file for any services that have been configured to be project specific.
+
+Org files (.org) live in both under ~/.cumulusci and in local project direcotries.
```
```suggestion
Org files (.org) live in both under ~/.cumulusci and in local project directories.
```
```suggestion
Org files (.org) live in both under ~/.cumulusci and in local project direcotries.
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -1,16 +1,40 @@
+DEFUALT_SERVICES.json file for any services that have been configured to be project specific.
+
+Org files (.org) live in both under ~/.cumulusci and in local project direcotries.
+The .org files store an encrypted access_token and ____________. There is a
```
```suggestion
The .org files store an encrypted access_token and other org attributes. There is a
```
```suggestion
The .org files store an encrypted access_token and ____________. There is a
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -26,107 +50,28 @@ def global_config_dir(self):
             return
-        full_path = os.path.join(dirname, filename)
+        full_path = Path(f"{dirname}/{filename}")
         if not os.path.exists(full_path):
```
```suggestion
        if not full_path.exists():
```
```suggestion
        if not os.path.exists(full_path):
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+            Path(self.project_local_dir) if project else Path(self.global_config_dir)
+        )
+        with open(dir_path / DEFAULT_SERVICES_FILENAME, "w") as f:
+            f.write(json.dumps(self._default_services))
```
Doesn't this mean that setting the default for one service type as a project-level default will turn the defaults for all other services into project-level defaults as well? I think when a default service is set at the project level we want to do a more targeted update of the project-level DEFAULT_SERVICES file to only add/update that one service in it.

Br4nd0R replied:
good catch

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+        the DEFAULT_SERVICES.json file in.
+        """
+        if dir_path is None:
+            return  # pragma: no cover
```
Can this actually happen?

Br4nd0R replied:
Just tested and looks like we can remove this.

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+        if not Path.is_dir(services_dir_path):
+            Path.mkdir(services_dir_path)
+
+        configured_service_types = self.project_config.config["services"].keys()
```
You don't actually need .keys() if you're just going to iterate over it. Iterating over a dict gives you the dict keys. Some people like to explicitly write .keys() though to make it clearer, and that's fine.

Br4nd0R replied:
I'll leave it for now. I think the variable name better conveys what the keys pertain to in this context.

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+        will be created.
+        """
+        services_dir_path = Path(f"{self.global_config_dir}/services")
+        if not Path.is_dir(services_dir_path):
+            Path.mkdir(services_dir_path)
```
```suggestion
        services_dir_path.mkdir(exist_ok=True)
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+        Scope is either 'project' or 'global' depending
+        on where the .service file is located."""
+        if not dir_path:
+            return  # pragma: no cover
```
Can this actually happen?

Br4nd0R replied:
Seems like it won't be an issue if we remove it.

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+
+        Default aliases are in the form `service_type__scope`.
+        Scope is either 'project' or 'global' depending
+        on where the .service file is located."""
```
I take it you decided not to try to support multiple projects with project-specific services by naming them based on the project name?

Br4nd0R replied:
This was something I just forgot to incorporate. I'll go back and name them based on project name.

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +105,300 @@ def unset_default_org(self):
+        raise ServiceNotConfigured(  # pragma: no cover
+            f"'{name}' service configuration could not be found. "
+            f"Maybe you need to run: cci service connect {name}"
+        )
```
I usually figure it's better to write the simple test for a function like this than to ignore coverage.

Br4nd0R commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -53,16 +118,12 @@ def _convert_connected_app(self):
         )
-        self.set_service("connected_app", ca_config)
+        # We're using an obnoxious alias to see who's using this
+        self.set_service("connected_app", "please_contact_sfdo_releng", ca_config)
```
@davisagli - Do you want me to just rip out the `_convert_connected_app()` function (and all references to it) in BaseProjectKeychain? (This would make me quite happy :) )

davisagli replied:
Yeah, I think we should go ahead and rip out everything that's related to connected apps as a thing that is separate from services, and just be sure to call it out in the release notes. This can be part of this PR or a new story for another sprint.

Br4nd0R commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -201,64 +206,136 @@ def get_org(self, name: str):
+        """Store a ServiceConfig in the keychain"""
+        if (
+            "services" not in self.project_config.config
+            or service_type not in self.project_config.config["services"]
```
I suppose it's more readable the way you mentioned. I was debugging tests and broke it out this way because of the context I was in at the time. I'll switch it back.

Br4nd0R commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -201,64 +206,136 @@ def get_org(self, name: str):
+    def _init_default_services(self):
+        pass
+
+    def set_service(self, service_type, alias, service_config, project=False):
```
@davisagli - Looks like this needs to be present as no other keychain class currently overrides this function. If we're going to merge EncryptedFileProjectKeychain with BaseEncryptedProjectKeychain, I could add a new `set_service()` method there that accepts the `project` kwarg.

davisagli replied:
@Br4nd0R I thought we decided all services would be stored at the global level and only defaults would be overridable at the project level. With that in mind, what's the purpose of the project flag here?

Br4nd0R replied:
You are correct; all services are stored at the global level. Since this is the only place where the `set_service()` method is defined, then when it is invoked from an instance of `EncryptedFileProjectKeychain` we still need to be able to pass in this argument so that the keychain knows that we need to write out the new project level default service into the local project's `DEFAULT_SERVICES.json` file in `EncryptedFileProjectKeychain._save_default_services()`.

But I think that we can remove the project argument from both `BaseEncryptedProjectKeychain` and `EncryptedFileProjectKeychain`'s `_set_encrypted_service()` methods.

davisagli replied:
I think if a new service is being set as the default because it's the first service of a type, we always want to set that at the global level, so don't need the project flag here. Explicitly setting a project-level override of the default can be done by a separate call to set_default_service.

Br4nd0R commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -201,64 +206,136 @@ def get_org(self, name: str):
+        if alias == default:
+            raise ServiceNotValid(
+                f"Service alias cannot be the default alias: {default}"
+            )
```
We will definitely allow people to adjust the names of services that have been migrated. This comment made me realize that this check is actually incorrect. This should be ensuring that when we set a new service that there is not another service of the same type with the same alias.

@codexmarcel will be implementing the `cci service rename <service_type> <current_alias> <new_alias>` command in this PR so renaming will be possible.

davisagli replied:
Are you sure we don't want to allow connecting a service if the name already exists? The use case I have in mind is _updating_ a service configuration. I tried doing this for an existing service and got this error. I think if the service name already exists we should make the `cci service connect` command prompt the user for confirmation, but still provide a way to replace it.

Br4nd0R commented on cumulusci/core/keychain/base_encrypted_project_keychain.py:
```diff
@@ -27,23 +31,37 @@ class BaseEncryptedProjectKeychain(BaseProjectKeychain):
+            self.services[service_type] = {}
+            # TODO: Should this be moved after _set_encrypted_service?
+            self._default_services[service_type] = alias
+            self._save_default_services(project)
```
I've made a new work item to address this in a future sprint: W-9110395

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -825,14 +844,19 @@ def callback(*args, **kwargs):
+            )
             if project:
-                click.echo(f"{name} is now configured for this project.")
+                click.echo(f"{service_type} is now configured for this project.")
```
This is potentially misleading. We are now only using the project flag to determine where the service gets set as the default, and the newly connected service will only get set as the default if it is the first one of this service_type.

I think we might want to do this:
1. Add a new `--default` flag which determines whether to set the new service as the global default for its type.
2. Repurpose `--project` to mean that the new service should be set as the project-specific default for its type.
3. Set both of these defaults via separate calls to set_default_service, so set_service doesn't need the project flag.

Br4nd0R replied:
> 3\. Set both of these defaults via separate calls to set_default_service, so set_service doesn't need the project flag.

set_service would still need the project flag it is going to set a default for the first service of a given type that is added. I suppose that I could take that logic out and implment it in the cci command. So if after we add a given service, and we detect that it is the first of its kind, then we call set_default_service() in that case as well (only if  --default and --project 
are not specified)

davisagli replied:
@Br4nd0R I think it's okay to keep the "set the first service of a type as the default for that type" logic in the keychain (and that seems preferable to make sure that it always happens). This implicit default should always be a global default, not project. And then the cci command can additionally use set_default_service to make it a project-level service if that's what the user asked for.

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -843,32 +867,63 @@ def service_connect():
         )
 
 
+@service.command(name="default", help="Show the details of a connected service")
```
wrong help

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -843,32 +867,63 @@ def service_connect():
+        click.echo(f"An error occurred setting the default service: {e}")
+        return
+
+    click.echo(f"'{service_name}' set as the default for {service_type} services.")
```
Doesn't this need a way to set a project-level override of the default?

davisagli replied:
Confirmed that works, but can we show a message that indicates which default was set (e.g. "for all projects" vs "for this project").

davisagli commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -24,46 +24,55 @@ class BaseProjectKeychain(BaseConfig):
+    def _validate_key(self):
         pass
 
+    def change_key(self, key):
```
This isn't used anywhere other than tests; we can probably remove it.

davisagli commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -201,64 +154,138 @@ def get_org(self, name: str):
+            not self.project_config.services
+            or service_type not in self.project_config.services
+        ):
+            self._raise_service_not_valid(service_type)
```
Should this check of the service type be part of _validate_service?

Br4nd0R replied:
I like it.

davisagli commented on cumulusci/core/keychain/base_project_keychain.py:
```diff
@@ -201,64 +154,138 @@ def get_org(self, name: str):
+        if alias == service_type:
+            raise ServiceNotValid(
+                "Service alias cannot be the same as the service type."
+            )
```
Is there a reason we need this restriction or is it just to avoid confusion?

Br4nd0R replied:
This was simply to avoid confusion.

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +89,410 @@ def unset_default_org(self):
+                f"No service of type {service_type} configured with name: {alias}"
+            )
+        elif alias == self._default_services[service_type]:
+            return  # already the default
```
But it might be the global default and user is trying to set it as the project-specific default (so that it won't change for this project if the global default is later changed)

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +89,410 @@ def unset_default_org(self):
+            )
+        elif current_alias not in self.services[service_type]:
+            raise ServiceNotConfigured(
+                f"No service of type {service_type} configrued with the name: {current_alias}"
```
```suggestion
                f"No service of type {service_type} configured with the name: {current_alias}"
```

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +89,410 @@ def unset_default_org(self):
+
+        self.services[service_type][new_alias] = self.services[service_type][
+            current_alias
+        ]
```
Don't we need a check to make sure the new_alias doesn't already exist?

Br4nd0R replied:
good catch

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +89,410 @@ def unset_default_org(self):
+        current_filepath.replace(new_filepath)
+
+        # look through all DEFAULT_SERVICE.json files and
+        # change current_alias to new_alias (if present)
```
Kudos for thinking of this

davisagli commented on docs/cli.rst:
```diff
@@ -263,6 +263,61 @@ and :doc:`Connect Persistent Orgs <connected_orgs>`.
+.. code-block:: console
+
+    $ cci service rename <service_type> <old_name> <new_name>
+
```
Have you checked for existing docs that mention services to see if they need to be updated?

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -825,14 +844,19 @@ def callback(*args, **kwargs):
-        return ret
+        params.append(click.Argument(["service_name"]))
+        return click.Command(service_type, params=params, callback=callback)
 
```
github won't let me comment on the line below where the "service connect" command is registered, but let's adjust the help to say "Connect an external service to CumulusCI" instead of "Connect a CumulusCI task service"

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -843,32 +867,63 @@ def service_connect():
 
 
 @service.command(name="info", help="Show the details of a connected service")
+@click.argument("service_type")
```
Should we make this optional so you can get the info for the default service without specifying its name?

davisagli commented on cumulusci/core/keychain/encrypted_file_project_keychain.py:
```diff
@@ -160,12 +89,410 @@ def unset_default_org(self):
+            return None
+
+        scope = "global" if item.parent.name == ".cumulusci" else "project"
+        new_filename = f"{service_type}__{scope}.service"
```
Should this just be f"{scope}.service"? I was expecting to have a github service named `global` after the migration, but instead I have one named `github__global`:

<img width="272" alt="Screen Shot 2021-04-09 at 11 02 58 AM" src="https://user-images.githubusercontent.com/49014/114200695-745ba400-9923-11eb-90c9-07a93f7ef9b2.png">

Br4nd0R replied:
We did intend to have the names like this but I think your suggestion is better.

davisagli commented:
This is only tangentially connected to this PR, but we should remove the service types `apextestsdb` and `saucelabs` from the universal cumulusci.yml, because we've already removed the code for the tasks that used them.

davisagli commented:
It can be a separate work item but we're going to want a `cci service remove <service_type> <service_name>` command.

Br4nd0R commented:
> This is only tangentially connected to this PR, but we should remove the service types `apextestsdb` and `saucelabs` from the universal cumulusci.yml, because we've already removed the code for the tasks that used them.

I went ahead and removed these from the standard library, and made a 1pt story for this and included it in the sprint.

davisagli commented on cumulusci/cli/cci.py:
```diff
@@ -796,25 +813,53 @@ def _build_param(self, attribute, details):
+            set_global_default = kwargs.pop("default", False)
+            if set_global_default and set_project_default:
+                raise click.UsageError(
+                    "Cannot specify both --default and --project. Please choose one."
```
This is fine, but there's no technical reason they need to be mutually exclusive.

Br4nd0R replied:
I switched it to accept both.
